File: clients/soa_services/parameter/python/src/crosslab/soa_services/parameter/parameter_service.py
import json
from typing import Dict, List, Union

from crosslab.soa_client.connection import Connection, DataChannel
from crosslab.soa_client.service import Service
from pyee.asyncio import AsyncIOEventEmitter

from crosslab.soa_services.parameter.messages import (
    ParameterChangedEvent,
    ParameterConfig,
    ParameterDescription,
    ParameterServiceConfig,
)


class ParameterService__Producer(Service):
    service_type = "https://api.goldi-labs.de/serviceTypes/parameter"
    service_direction = "producer"
    service_id: str
    parameter_descriptions: List[ParameterDescription]

    def __init__(
        self, service_id: str, parameter_descriptions: List[ParameterDescription]
    ):
        self.service_id = service_id
        self.parameter_descriptions = parameter_descriptions

    def getMeta(self):
        return {
            "serviceId": self.service_id,
            "serviceType": self.service_type,
            "serviceDirection": self.service_direction,
            "parameters": self.parameter_descriptions,
        }

    def setupConnection(
        self, connection: Connection, serviceConfig: ParameterServiceConfig
    ):
        self.channel = DataChannel()
        if connection.tiebreaker:
            connection.transmit(serviceConfig, "data", self.channel)
        else:
            connection.receive(serviceConfig, "data", self.channel)

    def teardownConnection(self, connection: Connection):
        pass

    async def updateParameter(self, parameter: str, value: float):
        self.channel.send(json.dumps({"parameter": parameter, "value": value}))


class Parameter(ParameterDescription):
    value: float


class ParameterService__Consumer(Service, AsyncIOEventEmitter):
    service_type = "https://api.goldi-labs.de/serviceTypes/parameter"
    service_direction = "consumer"
    service_id: str
    parameter_descriptions: List[ParameterDescription]

    parameters: Dict[Connection, List[ParameterConfig]]
    consolidated_parameters: List[ParameterDescription]

    def __init__(
        self, service_id: str, parameter_descriptions: List[ParameterDescription]
    ):
        AsyncIOEventEmitter.__init__(self)
        self.service_id = service_id
        self.parameter_descriptions = parameter_descriptions
        self.parameters = {}

    def getMeta(self):
        return {
            "serviceId": self.service_id,
            "serviceType": self.service_type,
            "serviceDirection": self.service_direction,
        }

    def setupConnection(
        self, connection: Connection, serviceConfig: ParameterServiceConfig
    ):
        self.channel = DataChannel()
        self.channel.on("data", lambda data: self.handleData(data))
        if connection.tiebreaker:
            connection.transmit(serviceConfig, "data", self.channel)
        else:
            connection.receive(serviceConfig, "data", self.channel)
        self.parameters[connection] = serviceConfig["parameters"]

    def teardownConnection(self, connection: Connection):
        pass

    def handleData(self, data: Union[str, bytes]):
        if isinstance(data, str):
            msg = json.loads(data)
            event: ParameterChangedEvent = {
                "parameter": msg["parameter"],
                "value": msg["value"],
            }
            self.emit("parameterChanged", event)

File: clients/soa_services/parameter/js/src/parameter_service.ts
import {
  Consumer,
  Producer,
  Service,
  ServiceConfiguration,
} from '@cross-lab-project/soa-client';
import { DataChannel } from '@cross-lab-project/soa-client';
import { PeerConnection } from '@cross-lab-project/soa-client';
import { TypedEmitter } from 'tiny-typed-emitter';

import {
  ParameterDescription,
  ParameterServiceEvent,
  ParameterServiceSetupEvent,
} from './messages';

type ServiceType = 'https://api.goldi-labs.de/serviceTypes/parameter';
const ServiceType: ServiceType = 'https://api.goldi-labs.de/serviceTypes/parameter';

export interface MessageServiceConfiguration extends ServiceConfiguration {
  serviceType: ServiceType;
  parameters: ParameterDescription[];
}
function checkConfig(
  config: ServiceConfiguration,
): asserts config is MessageServiceConfiguration {
  if (config.serviceType !== ServiceType) {
    //throw Error("Service Configuration needs to be for MessageService type");
  }
}

interface ParameterService__ProducerEvents {
  setup: (event: ParameterServiceSetupEvent) => void;
}

export class ParameterService__Producer
  extends TypedEmitter<ParameterService__ProducerEvents>
  implements Service
{
  serviceType = ServiceType;
  serviceId: string;
  serviceDirection = Producer;
  channel?: DataChannel;

  constructor(serviceId: string) {
    super();
    this.serviceId = serviceId;
  }

  getMeta() {
    return {
      serviceType: this.serviceType,
      serviceId: this.serviceId,
      serviceDirection: this.serviceDirection,
    };
  }

  setupConnection(connection: PeerConnection, serviceConfig: ServiceConfiguration): void {
    checkConfig(serviceConfig);
    const channel = new DataChannel();
    if (connection.tiebreaker) {
      connection.transmit(serviceConfig, 'data', channel);
    } else {
      connection.receive(serviceConfig, 'data', channel);
    }
    this.channel = channel;
    this.emit('setup', { parameters: serviceConfig.parameters });
  }

  async updateParameter(parameter: string, value: number) {
    if (this.channel !== undefined) {
      await this.channel.ready();
      this.channel.send(JSON.stringify({ parameter: parameter, value: value }));
    }
  }
}

interface ParameterService__ConsumerEvents {
  update: (event: ParameterServiceEvent) => void;
}

export class ParameterService__Consumer
  extends TypedEmitter<ParameterService__ConsumerEvents>
  implements Service
{
  serviceType = ServiceType;
  serviceId: string;
  serviceDirection = Consumer;
  parameterDescriptions: ParameterDescription[];

  constructor(serviceId: string, parametersDescriptions: ParameterDescription[]) {
    super();
    this.parameterDescriptions = parametersDescriptions;
    this.serviceId = serviceId;
  }

  getMeta() {
    return {
      serviceType: this.serviceType,
      serviceId: this.serviceId,
      serviceDirection: this.serviceDirection,
      parameters: this.parameterDescriptions,
    };
  }

  setupConnection(connection: PeerConnection, serviceConfig: ServiceConfiguration): void {
    checkConfig(serviceConfig);
    const channel = new DataChannel();
    channel.ondata = this.handleData.bind(this);
    if (connection.tiebreaker) {
      connection.transmit(serviceConfig, 'data', channel);
    } else {
      connection.receive(serviceConfig, 'data', channel);
    }
  }

  handleData(data: string | Blob | ArrayBuffer | ArrayBufferView) {
    if (typeof data === 'string') {
      const msg = JSON.parse(data);
      this.emit('update', { parameter: msg.parameter, value: msg.value });
    }
  }
}
